### PR TITLE
Replace HttpVfs entry cache to Caffeine cache

### PR DESCRIPTION
**Pull request by imasahiro**
Related to https://github.com/line/armeria/issues/440

TODOs:
* Use AsyncLoadingCache instead of LoadingCache
* ~~Expose cache stats~~

### CODE REVIEW

**Comment by anuraaga** [comment]
Can you clarify whether the TODOs are for future PRs - I guess so (or otherwise the title would be WIP) but want to confirm.

- I had looked into migrating to `AsyncLoadingCache` once a while ago and gave up since the `Vfs` abstractions seemed to deep to work well with it - just wondering what do you think?
- For metrics, we may want to also decide whether armeria is going prometheus-only, prometheus-first, or dual-prometheus/dropwizard always.

**Comment by imasahiro** [comment]
For AsyncLoadingCache, yeah, I also gave up to convert. So I would like to leave it for future PRs.

For metrics, I added prometheus metrics just now https://github.com/imasahiro/armeria/commit/a0157e9347e54cb4a01774cd827256781e190483 but I didn't decide my mind yet (prometheus only or both prometheus and dropwizard, etc.)

File: core/src/main/java/com/linecorp/armeria/server/file/HttpVfsCacheStatsCollector.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.linecorp.armeria.server.file;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpVfsCacheStatsCollector.java]: We need the license here~~

**Comment by trustin** [comment]
Could you fix the build failure?

**Comment by trustin** [comment]
Will merge once the build error is fixed.

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 2]
```diff
@@ -98,16 +100,18 @@ public static HttpFileService forVfs(HttpVfs vfs) {
...
         this.config = requireNonNull(config, "config");
 
         if (config.maxCacheEntries() != 0) {
-            cache = Collections.synchronizedMap(new LruMap<String, CachedEntry>(config.maxCacheEntries()));
+            cache = Caffeine.newBuilder()
+                            .maximumSize(config.maxCacheEntries())
+                            .recordStats()
+                            .build(this::getEntryWithoutCache);
+            CacheMetricsCollector collector = new CacheMetricsCollector().register();
+            collector.addCache(getClass().getSimpleName(), cache);
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: Q. I need to fill a unique cache name for each `HttpFileService` instance but what is suitable name for this?

Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: That's a good question. We could use the `PathMapping.metricName()` + '.vfsCache' or something?

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 3]
```diff
@@ -98,21 +105,33 @@ public static HttpFileService forVfs(HttpVfs vfs) {
 
     private final HttpFileServiceConfig config;
 
-    /**
-     * An LRU cache map that releases the buffer that contains the cached content.
-     */
-    private final Map<String, CachedEntry> cache;
+    private final LoadingCache<PathAndEncoding, CachedEntry> cache;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: nit: nullable?

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 4]
```diff
@@ -367,4 +384,32 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
             this.headerValue = headerValue;
         }
     }
+
+    private static final class PathAndEncoding {
+        private final String path;
+        @Nullable
+        private final String contentEncoding;
+
+        private PathAndEncoding(String path, @Nullable String contentEncoding) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: nit: we can remove the access modifier

**Comment by trustin** [comment]
The build is still failing.

**Comment by imasahiro** [comment]
Resolved build build.

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 5]
```diff
@@ -98,21 +102,35 @@ public static HttpFileService forVfs(HttpVfs vfs) {
...
         }
     }
 
+    @Override
+    public void serviceAdded(ServiceConfig cfg) throws Exception {
+        final MeterRegistry registry = cfg.server().meterRegistry();
+        if (cache != null) {
+            CaffeineMetricSupport.setup(
+                    registry,
+                    new MeterId("com.linecorp.armeria.server.file.vfsCache",
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: `armeria.server.file.vfsCache`?

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 6]
```diff
@@ -98,21 +102,35 @@ public static HttpFileService forVfs(HttpVfs vfs) {
...
+                                "name", config.vfs().toString(),
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: `"rootDir"`?

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 7]
```diff
@@ -98,21 +102,35 @@ public static HttpFileService forVfs(HttpVfs vfs) {
...
+                                "pathMapping", cfg.pathMapping().meterTag()),
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: How about adding:

    "virtualHost", cfg.virtualHost().hostnamePattern()

as well?

**Comment by trustin** [comment]
Thanks, @imasahiro !

**Pull request merged**